pallets/circuit/src/lib.rs

```diff
@@ -372,53 +374,39 @@ pub mod pallet {
                     bidding_timeouts_at <= &frame_system::Pallet::<T>::block_number()
                 })
                 .map(|(xtx_id, _bidding_timeouts_at)| {
-                    let mut local_ctx = Self::setup(
-                        CircuitStatus::PendingBidding,
-                        &Self::account_id(),
-                        Some(xtx_id),
-                    )
-                        .unwrap();
-
-                    // Ensure Circuit::PendingBidding status
-                    if local_ctx.xtx.status != CircuitStatus::PendingBidding {
-                        Self::kill(&mut local_ctx, CircuitStatus::DroppedAtBidding);
-                        return
-                    }
-
-                    let current_step = local_ctx.xtx.steps_cnt.0;
-                    for mut fsx in local_ctx.full_side_effects[current_step as usize].iter_mut() {
-                        let sfx_id = fsx.generate_id::<SystemHashing<T>, T>(local_ctx.xtx_id);
-                        if let Some(best_sfx_bid) = <PendingSFXBids<T>>::get(xtx_id, sfx_id) {
-                            fsx.best_bid = Some(best_sfx_bid);
-                        } else {
-                            // error - some FSX don't have bids
-                            Self::kill(&mut local_ctx, CircuitStatus::DroppedAtBidding);
+                    Machine::<T>::compile_infallible(
+                        &mut Machine::<T>::load_xtx(xtx_id).expect("xtx_id corresponds to a valid Xtx when reading from PendingXtxBidsTimeoutsMap storage"),
+                        |current_fsx, _local_state, _steps_cnt, status, _requester| {
+                            match status {
+                                CircuitStatus::PendingBidding | CircuitStatus::InBidding => {},
+                                _ => return PrecompileResult::TryKill(Cause::Timeout)
+                            }
+                            for mut fsx in current_fsx.iter_mut() {
+                                let sfx_id = fsx.generate_id::<SystemHashing<T>, T>(xtx_id);
+                                // Either assign best bid to FSX or Kill entire Xtx with DroppedAtBidding cause.
+                                if let Some(best_sfx_bid) = <PendingSFXBids<T>>::get(xtx_id, sfx_id)
+                                {
+                                    fsx.best_bid = Some(best_sfx_bid);
+                                } else {
+                                    // Error - at least one FSX has no bid
+                                    return PrecompileResult::TryKill(Cause::Timeout)
```

Comment:
Doesnt this lead to slashing again, even though we just refund on droppedAtBidding?